libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs

```diff
@@ -0,0 +1,707 @@
+  (Shelley _) -> Value p <$> arbitrary
+  (Allegra _) -> Value p <$> arbitrary
+  (Mary _) -> Value p <$> arbitrary
+  (Alonzo _) -> Value p <$> arbitrary
+  (Babbage _) -> Value p <$> arbitrary
+  (Conway _) -> Value p <$> arbitrary
+
+genTxOut :: Proof era -> Gen (TxOut era)
+genTxOut p = case p of
+  (Shelley _) -> TxOut p <$> arbitrary
+  (Allegra _) -> TxOut p <$> arbitrary
+  (Mary _) -> TxOut p <$> arbitrary
+  (Alonzo _) -> TxOut p <$> arbitrary
+  (Babbage _) -> TxOut p <$> arbitrary
+  (Conway _) -> TxOut p <$> arbitrary
+
+genPParams :: Proof era -> Gen (PParams era)
+genPParams p = case p of
+  (Shelley _) -> PParams p <$> arbitrary
+  (Allegra _) -> PParams p <$> arbitrary
+  (Mary _) -> PParams p <$> arbitrary
+  (Alonzo _) -> PParams p <$> arbitrary
+  (Babbage _) -> PParams p <$> arbitrary
+  (Conway _) -> PParams p <$> arbitrary
+
+genPParamsUpdate :: Proof era -> Gen (PParamsUpdate era)
+genPParamsUpdate p = case p of
+  (Shelley _) -> PParamsUpdate p <$> genShelleyPParamsUpdate defaultConstants def
+  (Allegra _) -> PParamsUpdate p <$> genShelleyPParamsUpdate defaultConstants def
+  (Mary _) -> PParamsUpdate p <$> genShelleyPParamsUpdate defaultConstants def
+  (Alonzo _) -> PParamsUpdate p <$> arbitrary
+  (Babbage _) -> PParamsUpdate p <$> arbitrary
+  (Conway _) -> PParamsUpdate p <$> arbitrary
+
+genProposedPPUpdates :: Proof era -> Gen (ProposedPPUpdates era)
+genProposedPPUpdates p = case p of
+  (Shelley _) -> ProposedPPUpdates p . Core.ProposedPPUpdates <$> arbitrary
+  (Allegra _) -> ProposedPPUpdates p . Core.ProposedPPUpdates <$> arbitrary
+  (Mary _) -> ProposedPPUpdates p . Core.ProposedPPUpdates <$> arbitrary
+  (Alonzo _) -> ProposedPPUpdates p . Core.ProposedPPUpdates <$> arbitrary
+  (Babbage _) -> ProposedPPUpdates p . Core.ProposedPPUpdates <$> arbitrary
+  (Conway _) -> ProposedPPUpdates p . Core.ProposedPPUpdates <$> arbitrary
+
+genGovernanceState :: Proof era -> Gen (GovernanceState era)
+genGovernanceState p = case p of
+  (Shelley _) -> GovernanceState p <$> arbitrary
+  (Allegra _) -> GovernanceState p <$> arbitrary
+  (Mary _) -> GovernanceState p <$> arbitrary
+  (Alonzo _) -> GovernanceState p <$> arbitrary
+  (Babbage _) -> GovernanceState p <$> arbitrary
+  (Conway _) -> pure $ GovernanceState p Core.emptyGovernanceState
+
+genUTxO :: Proof era -> Gen (UTxO era)
+genUTxO p = case p of
+  (Shelley _) -> arbitrary
+  (Allegra _) -> arbitrary
+  (Mary _) -> arbitrary
+  (Alonzo _) -> arbitrary
+  (Babbage _) -> arbitrary
+  (Conway _) -> arbitrary
+
+-- =========================================================================
+-- SumsTo x <= [One y, SumMap z]
+--          ^ paramerterize over the condition
+-- EQL = (==), LTH = (<), LTE = (<=), GTH = (>), GTE = (>=)
+
+data SumCond = EQL | LTH | LTE | GTH | GTE | CondNever [String] | CondAny
+  deriving (Eq)
+
+negSumCond :: SumCond -> SumCond
+negSumCond EQL = EQL
+negSumCond LTH = GTH
+negSumCond LTE = GTH
+negSumCond GTH = LTH
```

Comment:
Shouldn't negation be like this: `not (x < y) == x >= y` and `not (x > y) == x <= y`?
```suggestion
negSumCond LTH = GTE
negSumCond LTE = GTH
negSumCond GTH = LTE
```

Reply:
This is more subtle than that, because the domain is just the positive numbers, and not every number has a predecessor. So this transformation is sound, but not complete.

Reply:
That sounds good. Can you leave a short comment about it for the future reader? Cause it looks wrong, and lonesome in the future might try to wrongfully "fix" it.

Reply:
Ok, we now call it reverseOrdCond, give an example of why it is needed, and give a ink to a URL where they explain how it arises.